File: app/src/main/java/br/com/jonathanzanella/myexpenses/source/SourcePresenter.java
package br.com.jonathanzanella.myexpenses.source;

import android.os.AsyncTask;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.exceptions.InvalidMethodCallException;
import br.com.jonathanzanella.myexpenses.validations.OperationResult;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;

class SourcePresenter {
	private final SourceRepository repository;
	private SourceContract.View view;
	@Nullable
	private SourceContract.EditView editView;
	private Source source;

	SourcePresenter(SourceRepository repository) {
		this.repository = repository;
	}

	void attachView(SourceContract.View view) {
		this.view = view;
	}

	void attachView(SourceContract.EditView view) {
		this.view = view;
		this.editView = view;
	}

	void detachView() {
		this.view = null;
		this.editView = null;
	}

	void viewUpdated() {
		if (source != null) {
			if (editView != null) {
				editView.setTitle(R.string.edit_source_title);
			} else {
				String title = view.getContext().getString(R.string.source);
				view.setTitle(title.concat(" ").concat(source.getName()));
			}
			view.showSource(source);
		} else {
			if(editView != null)
				editView.setTitle(R.string.new_source_title);
		}
	}

	@UiThread
	void loadSource(final String uuid) {
		new AsyncTask<Void, Void, Void>() {

			@Override
			protected Void doInBackground(Void... voids) {
				source = repository.find(uuid);
				return null;
			}

			@Override
			protected void onPostExecute(Void aVoid) {
				super.onPostExecute(aVoid);
				viewUpdated();
			}
		}.execute();
	}

	void save() {
		if(editView == null)
			throw new InvalidMethodCallException("save", getClass().toString(), "View should be a Edit View");
		if(source == null)
			source = new Source();
		source = editView.fillSource(source);
		OperationResult result = repository.save(source);

		if(result.isValid()) {
			editView.finishView();
		} else {
			for (ValidationError validationError : result.getErrors())
				editView.showError(validationError);
		}
	}

	String getUuid() {
		return source != null ? source.getUuid() : null;
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/source/SourcePresenter.kt
package br.com.jonathanzanella.myexpenses.source

import android.os.AsyncTask
import android.support.annotation.UiThread
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.exceptions.InvalidMethodCallException
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread

class SourcePresenter(private val repository: SourceRepository) {
    private var view: SourceContract.View? = null
    private var editView: SourceContract.EditView? = null
    private var source: Source? = null

    fun attachView(view: SourceContract.View) {
        this.view = view
    }

    fun attachView(view: SourceContract.EditView) {
        this.view = view
        this.editView = view
    }

    fun detachView() {
        this.view = null
        this.editView = null
    }

    fun viewUpdated() {
        val s = source
        if (s != null) {
            val v = editView
            if (v != null) {
                v.setTitle(R.string.edit_source_title)
            } else {
                view!!.let {
                    val title = it.context.getString(R.string.source)
                    it.setTitle(title + " " + s.name)
                }
            }
            view!!.showSource(s)
        } else {
            editView?.setTitle(R.string.new_source_title)
        }
    }

    @UiThread
    fun loadSource(uuid: String) {
        object : AsyncTask<Void, Void, Void>() {

            override fun doInBackground(vararg voids: Void): Void? {
                source = repository.find(uuid)
                return null
            }

            override fun onPostExecute(aVoid: Void?) {
                super.onPostExecute(aVoid)
                viewUpdated()
            }
        }.execute()
    }

    @UiThread
    fun save() {
        val v = editView ?: throw InvalidMethodCallException("save", javaClass.toString(), "View should be a Edit View")
        source = v.fillSource(source ?: Source())
        doAsync {
            val result = repository.save(source!!)

            uiThread {
                if (result.isValid) {
                    v.finishView()
                } else {
                    for (validationError in result.errors)
                        v.showError(validationError)
                }
            }
        }
    }

    val uuid: String?
        get() = source?.uuid
}
